### PR TITLE
build: create merge script w/ cherry-picking 

### PR DESCRIPTION
**Pull request by devversion**
Creates a merge script that automatically merges pull
requests from Github and cherry-picks them into the
right publish branches based on the PR target label.

The script has been designed in an extensible and programmatically
acessible way, so that it can be used in other Angular repositories too.

The script provides infrastructure for PR merge safety checks. e.g.
it currently enforces that the CLA is signed, or that the merge ready
label is assigned, or that no CI checks are failing / pending.

### CODE REVIEW

File: scripts/merge-script/config.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    errors.push('Repository configuration needs to specify a `user` and repository `name`.');
+  }
+  if (!config.claSignedLabel) {
+    errors.push('No CLA signed label configured.');
+  }
+  if (!config.mergeReadyLabel) {
+    errors.push('No merge ready label configured.');
+  }
+  return errors;
+}
```
Comment (devversion) [inline scripts/merge-script/config.ts]: The config parsing will be easy to integrate w` `getDevInfraConfig()` if we eventually move it to `/dev-infra`.

File: scripts/merge-script/cli.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ * into branches based on the `PR target` labels that have been set in a configuration. The script
+ * aims to reduce the manual work that needs to be performed to cherry-pick a PR into multiple
+ * branches based on a target label.
+ */
+async function main() {
+  const {config, pullRequests, force, dryRun, githubToken: cliGithubToken} = parseCommandLine();
+  const githubToken = cliGithubToken || process.env.GITHUB_TOKEN || process.env.TOKEN;
+
+  if (!githubToken) {
+    console.error(
```
Comment (josephperrott) [inline scripts/merge-script/cli.ts]: I wonder if it would be worth creating a common util that contains `error()`, `info()` and `log()`, that perform `console.error`, `console.info`, etc. with the `chalk` colors already applied?   We are going to have to do a ton of logging things like this throughout our tooling and consistency would be nice.

If thats something that sounds like a good idea to you, I would also throw out the suggestion of these functions taking in multiple args, and placing each arg on its own console line.

It would allow us to have this block become, IMO, more readable as
```js
if (!githubToken) {
  logging.error(
    'No Github token is set. Please set the `GITHUB_TOKEN` environment variable.',
    'Alternatively, pass the `--github-token` command line flag.');
  process.exit(1);
}
```

File: scripts/merge-script/cli.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+   * Handles the merge result by printing console messages, exiting the process
+   * based on the result, or by restarting the merge if force mode has been enabled.
+   */
+  async function handleMergeResult(
+      {failures, status, merged}: MergeResult, disableForceMergePrompt = false) {
+    const canForciblyMerge = failures ?
+        failures.every(failure => FORCE_OVERRIDABLE_FAILURES.includes(failure.status)) :
+        false;
+
+    if (status === MergeStatus.SUCCESS) {
```
Comment (josephperrott) [inline scripts/merge-script/cli.ts]: I personally find this more readable, with `if` blocks that have returns since we are also either exiting the function or the entire process on each match.

File: merge-config.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+module.exports = () => {
```
[the remark this answers is not in the document]
Comment (devversion) [reply merge-config.js]: Yeah, we most likely will. I was thinking about making it support a JS file when this is moved over. Do you have any concerns with that?

File: scripts/merge-script/cli.ts (hunk 3 above)
Comment (devversion) [reply scripts/merge-script/cli.ts]: I converted it to a `switch` statement. Let me know if that is better.

File: scripts/merge-script/cli.ts (hunk 2 above)
Comment (devversion) [reply scripts/merge-script/cli.ts]: I like the idea of having a shared logging util. The only thing that concerns me is that usually the chalk statements are needed in the middle of a message. e.g `chalk.red(<..> ${chalk.bold(...)} <..>)`. That might be inconsistent then with the mid-sentence color changes.

Regardless of that though, I think that it would make sense to explore this. Potentially we could also have a `debug` method then (that is based on an environment variable).

Should we do that separately in the `dev-infra/` package?

File: scripts/merge-script/cli.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+/**
+ * Entry-point for the merge script CLI. The script can be used to merge multiple pull requests
+ * into branches based on the `PR target` labels that have been set in a configuration. The script
+ * aims to reduce the manual work that needs to be performed to cherry-pick a PR into multiple
+ * branches based on a target label.
+ */
+async function main() {
+  const {config, pullRequests, force, dryRun, githubToken: cliGithubToken} = parseCommandLine();
+  const githubToken = cliGithubToken || process.env.GITHUB_TOKEN || process.env.TOKEN;
```
[the remark this answers is not in the document]
Comment (devversion) [reply scripts/merge-script/cli.ts]: Yep. Changed!

File: scripts/merge-script/cli.ts (hunk 2 above)
Comment (josephperrott) [reply scripts/merge-script/cli.ts]: Yeah, I think its probably worth waiting and just doing it in the dev-infra package.

File: scripts/merge-script/index.ts [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+          // pull request merges as a Git thinks a cherry-pick is still in progress.
+          this._gitGraceful(['cherry-pick', '--abort']);
+          return {prNumber, status: PullRequestStatus.CHERRY_PICK_CONFLICTS, context: branch};
+        }
+      }
+      return null;
+    } catch {
+      return {prNumber, status: PullRequestStatus.UNKNOWN_MERGE_ERROR};
+    } finally {
+      // Always try to restore the branch if possible. We don't want to leave the
```
Comment (josephperrott) [inline scripts/merge-script/index.ts]: Additional to always trying to restore the branch we came from, we should also clean up all of the branches we created correct?

File: scripts/merge-script/index.ts [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+      const {data: {state, statuses}} = await this._github.repos.getCombinedStatusForRef(
+          {...this._repoParams, ref: prData.head.sha});
+
+      if (state === 'failure' && !force) {
+        failures.push({prNumber, status: PullRequestStatus.FAILING_CI_JOBS});
+        continue;
+      }
+      // The state is set to `pending` if a pull request does not have any status checks.
+      // https://developer.github.com/v3/repos/statuses/#get-the-combined-status-for-a-specific-ref
+      if (state === 'pending' && statuses.length !== 0 && !force) {
```
[the remark this answers is not in the document]
Comment (devversion) [reply scripts/merge-script/index.ts]: That's a good point. Done.

File: scripts/merge-script/index.ts [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+      fetchRefspecs.push(`refs/heads/${targetBranch}:${localTargetBranch}`);
+      pushRefspecs.push(`${localTargetBranch}:refs/heads/${targetBranch}`);
+    }
+
+    if (this._gitGraceful(['fetch', '-f', this._repoGitUrl, ...fetchRefspecs]).status !== 0) {
+      return {status: MergeStatus.UNKNOWN_GIT_ERROR};
+    }
+
+    for (const pullRequest of queue) {
+      const failure = await this._mergePullRequest(pullRequest);
```
[the remark this answers is not in the document]
Comment (devversion) [reply scripts/merge-script/index.ts]: Yeah. Done!

File: scripts/merge-script/cli.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        }
+        return false;
+      default:
+        throw Error(`Unexpected merge result: ${status}`);
+    }
+  }
+}
+
+/** Parses the command line and returns the passed options. */
+function parseCommandLine(): {
```
Comment (mmalerba) [inline scripts/merge-script/cli.ts]: Should we use some kind of command line parsing library? I'm always tempted to write stuff like this in Python since they have really nice parsing built in

Comment (devversion) [reply scripts/merge-script/cli.ts]: Yeah agreed! The plan is to use `yargs` for this if we move it over to `dev-infra/` in `angular/angular`. We decided to do that as a follow-up.

Comment (jelbourn) [reply scripts/merge-script/cli.ts]: Add a TODO?

File: scripts/merge-script/index.ts [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+  }
+
+  /**
+   * Merges the specified pull request. Method requires the temporary target branch
+   * to be fetched already. We don't want to fetch the target branches per pull request
+   * merge as it causes unnecessary fetch requests.
+   * @returns A pull request failure or null in case of success.
+   */
+  private async _mergePullRequest(
+      {prNumber, targetBranches, requiredBaseSha, commitsAmount}: PullRequest,
```
Comment (jelbourn) [inline scripts/merge-script/index.ts]: Could this destructuring be done in the method instead of in the definition? I just think it looks cleaner that way.

File: scripts/merge-script/index.ts [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+    };
+  }
+
+  /**
+   * Merges the specified pull request. Method requires the temporary target branch
+   * to be fetched already. We don't want to fetch the target branches per pull request
+   * merge as it causes unnecessary fetch requests.
+   * @returns A pull request failure or null in case of success.
+   */
+  private async _mergePullRequest(
```
Comment (jelbourn) [inline scripts/merge-script/index.ts]: It's not immediately clear what the difference between `merge` and `_mergePullRequest` is from the name. Maybe use names like `mergeAll` and `mergeSingle`?

File: scripts/merge-script/index.ts [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+      {prNumber, targetBranches, requiredBaseSha, commitsAmount}: PullRequest,
+      dryRun: boolean): Promise<PullRequestFailure|null> {
+    let previousBranch: string|null = null;
+    try {
```
Comment (jelbourn) [inline scripts/merge-script/index.ts]: Add a comment about which error(s) you're expecting with this `try`

File: scripts/merge-script/commit-message-filter.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+  process.stdin.on('end', () => {
+    console.info(rewriteCommitMessage(commitMessage, prNumber));
+  });
+}
+
+function rewriteCommitMessage(message, prNumber) {
+  const lines = message.split(/\n/);
+  // Add the pull request number to the commit message title. This matches what
+  // Github does when PRs are merged on the web through the `Squash and Merge` button.
+  lines[0] += ' (#' + prNumber + ')';
```
Comment (jelbourn) [inline scripts/merge-script/commit-message-filter.js]: Use template strings?

File: scripts/merge-script/config.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+  labels: TargetLabel[];
+  /** Pattern that matches labels which imply a signed CLA. */
+  claSignedLabel: string|RegExp;
+  /** Pattern that matches label which imply a merge ready pull request. */
+  mergeReadyLabel: string|RegExp;
+}
+
+/** Reads and validates the configuration file at the specified location. */
+export function readAndValidateConfig(filePath: string): {config?: Config, errors?: string[]} {
+  try {
```
Comment (jelbourn) [inline scripts/merge-script/config.ts]: Add comment about the type of error(s) you're expecting with this `try`

File: scripts/merge-script/console.ts [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    case PullRequestStatus.FAILING_CI_JOBS:
+      return `Failing CI jobs.`;
+    case PullRequestStatus.PENDING_CI_JOBS:
+      return `Pending CI jobs.`;
+    case PullRequestStatus.NO_TARGET_LABEL:
+      return `No target branch could be determined. Please ensure a target label is set.`;
+    case PullRequestStatus.NOT_MERGE_READY:
+      return `Not marked as merge ready.`;
+    case PullRequestStatus.CHERRY_PICK_CONFLICTS:
+      return `Could not cherry-pick pull request into branch: ${context}`;
```
Comment (jelbourn) [inline scripts/merge-script/console.ts]: Probably want to use the phrase "merge conflict" in this message; the current wording omits the reason the change can't be cherry-picked

File: scripts/merge-script/index.ts [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+      previousBranch = this._git(['rev-parse', '--abbrev-ref', 'HEAD']).stdout.trim();
```
Comment (jelbourn) [inline scripts/merge-script/index.ts]: I'd add a comment like "Remember the branch we're on now so we can change back to it at the end"

File: scripts/merge-script/index.ts (hunk 10 above)
Comment (devversion) [reply scripts/merge-script/index.ts]: Yeah, that's a good point.

File: scripts/merge-script/index.ts (hunk 11 above)
Comment (devversion) [reply scripts/merge-script/index.ts]: Agreed. I'll try to make this more obvious.

File: scripts/merge-script/commit-message-filter.js (hunk 13 above)
Comment (devversion) [reply scripts/merge-script/commit-message-filter.js]: Yep. I forgot that when I copied the script from `angular/angular`.

**Comment by devversion** [comment]
@jelbourn Addressed your feedback (and also simplified things a bit). Also, are you aware of the fact that this script doesn't preserve the purple `Merged` state for pull requests? I investigated whether we can merge purely with the Github API, but that's not possible. Additionally, we can't merge using the API, and then cherry-pick into other branches as that would mean that autosquashing doesn't work. Autosquashing is commonly used in framework.

I think we should just acknowledge that this is a Github limitation. Github should just provide an API for changing the PR state to `merged`. Clearly merges using the `Squash & Merge` button aren't fast-forward either.. so Github is already tricking here.

**Comment by jelbourn** [comment]
Ah, if we can't preserve the purple "merged" status, then I would prefer to continue pressing the button in the UI. Maybe then the script could handle the cherry-picking after merging the PR.

**Comment by devversion** [comment]
@jelbourn I think the problem here is that this script should be also usable in other repositories. If it would be only used in the components repository, then we could definitely use the Github PR merge API, and perform the cherry-picking locally. 

Though I wanted the script to be part of the shared dev-infra tools, so that it can be used in framework too. Framework relies on auto-squashing which isn't supported by the Github merge buttons. We don't perform fast-forward merges (we cannot), so if we ever want that behavior, Github needs to provide an API for us to set the merged state.

Alternatively we can add opt-in options to the script that allow:

*  Squash merging through pressing the button in the UI. The script only does the cherry-picking. This would also simplify the process of doing manual commit message cleanup.
* Squash merging through the API (automatically). The script does the cherry-picking as needed.

Basically, the script would be made "suitable" for requirements from both repositories where the script is used. I personally would prefer having a single canonical way, but I understand the reason for having it show up as `Merged`. At the same time though.. I'd like to start using `fixup!` commits in the components repo, and not having to request for preserving of multiple commits (if needed).

Cherry-picking a merged PR through the Github API/UI also will complicate things, as it will be difficult for the script to find corresponding PR commits if merged through the `Rebase and Merge` button.

**Comment by devversion** [comment]
@jelbourn @josephperrott I made the changes we discussed in our meeting about this. The merge strategy for using the Github API turned a bit more code than initially anticipated, but I think that amount of code is reasonable and acceptable. Easy to remove if we ever want a canonical merge strategy across all repositories.

File: scripts/merge-script/config.ts [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+  /**
+   * Whether pull requests should be merged using the Github API. This can be enabled
+   * if projects want to have their pull requests show up as `Merged` in the Github UI.
+   * The downside is that fixup or squash commits no longer work as the Github API does
+   * not support this.
+   */
+  githubApiMerge?: {
+    default: GithubApiMergeMethod,
+    labels?: {pattern: string, method: GithubApiMergeMethod}[],
+  };
```
Comment (josephperrott) [inline scripts/merge-script/config.ts:57]: Can we require this is defined so its explicit which is being chosen?

```suggestion
  githubApiMerge: false | {
    default: GithubApiMergeMethod,
    labels?: {pattern: string, method: GithubApiMergeMethod}[],
  };
```

File: scripts/merge-script/index.ts [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+      // Always try to restore the branch if possible. We don't want to leave
+      // the repository in a different state than before.
+      if (previousBranch !== null) {
+        this.git.runGraceful(['checkout', '-f', previousBranch]);
+      }
+    }
+  }
+
+  /** Gets whether the current Git repository has uncommitted changes. */
+  private _hasUncommittedChanges(): boolean {
```
Comment (josephperrott) [inline scripts/merge-script/index.ts:117]: Does it make more sense for this to be a method on `GitClient`?

File: scripts/merge-script/index.ts [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+      return {status: MergeStatus.DIRTY_WORKING_DIR};
+    }
+
+    const pullRequestOrFailure = await loadAndValidatePullRequest(this, prNumber, force);
+
+    if (!isPullRequest(pullRequestOrFailure)) {
+      return {status: MergeStatus.FAILED, failure: pullRequestOrFailure};
+    }
+
+    const pullRequest: PullRequest = pullRequestOrFailure;
```
Comment (josephperrott) [inline scripts/merge-script/index.ts:69]: Is it possible for this to be done with a type guard?

`isPullRequest` can confirm the type of `pullRequestOrFailure`?

That would allow this to be:

```suggestion
    const pullRequest = await loadAndValidatePullRequest(this, prNumber, force);

    if (!isPullRequest(pullRequest)) {
      return {status: MergeStatus.FAILED, failure: pullRequestOrFailure};
    }
```

File: scripts/merge-script/pull-request.ts [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+  prNumber: number;
+  /** Title of the pull request. */
+  title: string;
+  /** Labels applied to the pull request. */
+  labels: string[];
+  /** List of branches this PR should be merged into. */
+  targetBranches: string[];
+  /** Branch that the PR targets in the Github UI. */
+  githubTargetBranch: string;
+  /** Amount of commits this pull request contains. */
```
Comment (josephperrott) [inline scripts/merge-script/pull-request.ts:41]: nit, can we do this as `commitCount`?

File: scripts/merge-script/pull-request.ts [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+/** Failure for pull request. */
+export interface PullRequestFailure {
+  /** Status of the associated pull request. */
+  status: PullRequestStatus;
+  /**
+   * Optional context for the pull request failure. e.g. Name of the branch that
+   * the pull request could not be cherry-picked into.
+   */
+  context?: any;
```
[the remark this answers is not in the document]
Comment (josephperrott) [reply scripts/merge-script/pull-request.ts:57]: I think to do so, we would need to overload this interface for each of the potential `PullRequestStatus`es, I am not entirely should how much value it would add.  But could definitely be considered.

File: scripts/merge-script/strategies/api-merge.ts [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+  async merge(pullRequest: PullRequest): Promise<PullRequestFailure|null> {
+    const {githubTargetBranch, prNumber, targetBranches, requiredBaseSha, needsCommitMessageFixup} =
+        pullRequest;
+    // If the pull request does not have its base branch set to any determined target
+    // branch, we cannot merge using the API.
+    if (targetBranches.every(t => t !== githubTargetBranch)) {
+      return {status: PullRequestStatus.MISMATCHING_TARGET_BRANCH, context: targetBranches};
+    }
+
+    // In case a required base is specified for this pull request, check if the pull
```
Comment (josephperrott) [inline scripts/merge-script/strategies/api-merge.ts:40]: `// In cases where a required base`?

File: scripts/merge-script/strategies/api-merge.ts [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+    // request contains the given commit. If not, return a pull request failure. This
+    // check is useful for enforcing that PRs are rebased on top of a given commit. e.g.
+    // a commit that changes the codeowner ship validation. PRs which are not rebased
```
Comment (josephperrott) [inline scripts/merge-script/strategies/api-merge.ts:43]: s/codeowner ship/code ownership

File: scripts/merge-script/strategies/api-merge.ts [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+    const method = this._getMergeActionFromPullRequest(pullRequest);
+    const cherryPickTargetBranches = targetBranches.filter(b => b !== githubTargetBranch);
+
+    // First cherry-pick the PR into all local target branches in dry-run mode. This is
+    // purely for testing so that we can figure out whether the PR can be cherry-picked
+    // into the other target branches. We don't want to merge the PR through the API, and
+    // then run into cherry-pick conflicts after the initial merge already completed.
+    const failure = await this._checkMergability(pullRequest, cherryPickTargetBranches);
+
+    // If the PR could not be cherry-picked into all  locally, then we know it can't be done
```
Comment (josephperrott) [inline scripts/merge-script/strategies/api-merge.ts:58]: `into all target branches locally`

File: scripts/merge-script/strategies/api-merge.ts [hunk 25]
```diff
@@ -0,0 +1,191 @@
...
+    const mergeOptions: PullsMergeParams = {
+      pull_number: prNumber,
+      merge_method: method,
+      ...this.git.repoParams,
+    };
+
+    // If a commit message fixup is needed and we are merging using squash, we prompt
+    // for commit message changes. Note that this does not work with other merge methods
+    // as the Github API only allows commit message modifications for squash merging.
+    if (needsCommitMessageFixup && method === 'squash') {
```
Comment (josephperrott) [inline scripts/merge-script/strategies/api-merge.ts:73]: Should we error in cases in which something is marked as `needsCommitMessageFixup` but is a `method` that is not `squash`?

File: scripts/merge-script/strategies/api-merge.ts [hunk 26]
```diff
@@ -0,0 +1,191 @@
...
+
+    // Cherry pick the merged commits into the remaining target branches.
+    const failedBranches = await this.cherryPickIntoTargetBranches(
+        `${sha}~${targetCommitsAmount}..${sha}`, cherryPickTargetBranches);
+
+    if (failedBranches.length) {
+      return {status: PullRequestStatus.MERGE_CONFLICTS, context: failedBranches};
+    }
+
+    this.pushTargetBranchesUpstream(cherryPickTargetBranches);
```
Comment (josephperrott) [inline scripts/merge-script/strategies/api-merge.ts:112]: Should we confirm the pushes are successful?  What happens if it were to fail here?

File: scripts/merge-script/strategies/api-merge.ts [hunk 27]
```diff
@@ -0,0 +1,191 @@
...
+   */
+  async _promptCommitMessageEdit(pullRequest: PullRequest, mergeOptions: PullsMergeParams) {
+    const commitMessage = await this._getDefaultSquashCommitMessage(pullRequest);
+    const {result} = await prompt<{result: string}>({
+      type: 'editor',
+      name: 'result',
+      message: 'Please update the commit message',
+      default: commitMessage,
+    });
+
```
Comment (josephperrott) [inline scripts/merge-script/strategies/api-merge.ts:129]: Should we also be validating the commit message that is created here?

File: scripts/merge-script/strategies/strategy.ts [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Config} from '../config';
+import {GitClient} from '../git';
+import {PullRequest, PullRequestFailure} from '../pull-request';
+
+/** Name of a temporary branch that contains the head of a currently-processed PR. */
+export const TEMP_PR_HEAD_BRANCH = 'merge_pr_head';
```
Comment (josephperrott) [inline scripts/merge-script/strategies/strategy.ts:14]: Should this branch name be unique?  Or is it okay that it gets reused?

File: scripts/merge-script/cli.ts [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+ * branches based on a target label.
+ */
+async function main() {
+  const {config, prNumber, force, githubToken} = parseCommandLine();
+
+  const api = new PullRequestMergeTask(config, githubToken);
+
+  // Merge the pull request and handle the result. If failures have been reported, exit
+  // the process with a non-zero exit code.
+  if (!await handleMergeResult(await api.merge(prNumber, false), force)) {
```
[the remark this answers is not in the document]
Comment (devversion) [reply scripts/merge-script/cli.ts:40]: 👍 Done.

File: scripts/merge-script/cli.ts [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+    }
+  }
+}
+
+// TODO(devversion): Use Yargs for this once the script has been moved to `angular/angular`.
+/** Parses the command line and returns the passed options. */
+function parseCommandLine():
+    {config: Config, force: boolean, prNumber: number, dryRun?: boolean, githubToken: string} {
+  const {config: configPath, githubToken: githubTokenArg, force, _: [prNumber]} =
+      minimist<any>(process.argv.slice(2), {
```
[the remark this answers is not in the document]
Comment (devversion) [reply scripts/merge-script/cli.ts:103]: Unfortunately not. We'll clean this up anyway in a follow-up when we switch to `yargs`.

File: scripts/merge-script/config.ts (hunk 17 above)
Comment (devversion) [reply scripts/merge-script/config.ts:57]: Yep. Done

File: scripts/merge-script/index.ts (hunk 18 above)
Comment (devversion) [reply scripts/merge-script/index.ts:117]: Yeah, I think that makes sense. Done.

File: scripts/merge-script/index.ts (hunk 19 above)
Comment (devversion) [reply scripts/merge-script/index.ts:69]: `isPullRequest` already is a type guard. The only difference I see here is that you renamed the variable from `pullRequestOrFailure` to `pullRequest`. I'm fine with that.

File: scripts/merge-script/pull-request.ts (hunk 20 above)
Comment (devversion) [reply scripts/merge-script/pull-request.ts:41]: Haha 😂 You always get me with that one. Sure, will change it!

File: scripts/merge-script/strategies/api-merge.ts (hunk 25 above)
Comment (devversion) [reply scripts/merge-script/strategies/api-merge.ts:73]: I like that.

File: scripts/merge-script/pull-request.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (devversion) [reply scripts/merge-script/pull-request.ts:57]: Yeah, I agree that this is not really ideal. My goal here is that the actual merge logic does not have to deal with writing messages (also since that avoids duplication).

I'm changing it to an abstract class with static methods on it. e.g.

```ts
return PullRequestFailure.mergeConflicts(failedBranches)
```

Let me know if that works for you all.

File: scripts/merge-script/strategies/strategy.ts (hunk 28 above)
Comment (devversion) [reply scripts/merge-script/strategies/strategy.ts:14]: It can be any arbitrary branch name. This is the one from the framework's merge script.

Comment (devversion) [reply scripts/merge-script/strategies/strategy.ts:14]: Tried to make this more clear in the description.

File: scripts/merge-script/strategies/api-merge.ts (hunk 26 above)
Comment (devversion) [reply scripts/merge-script/strategies/api-merge.ts:112]: If the push fails, then the script exits and the failure is printed out (git stderr). I don't think there is much we can do. This probably can happen if multiple PRs are merged at the same time. In those cases, the Git stderr message should explain this, and it's up to the consumer of this script to re-run (if possible).

File: scripts/merge-script/strategies/api-merge.ts (hunk 25 above)
Comment (devversion) [reply scripts/merge-script/strategies/api-merge.ts:73]: done.

File: scripts/merge-script/strategies/api-merge.ts (hunk 27 above)
Comment (devversion) [reply scripts/merge-script/strategies/api-merge.ts:149]: I'd say yes, but that seems to be something we can tackle as a follow-up? Especially since we are not having commit message validation in the components repo yet. I'd like to set it up soon, but I think we need some upstream changes to account for a different `scope` that we support.

**Comment by devversion** [comment]
@jelbourn @josephperrott Addressed most of the feedback. Please have another look.

File: scripts/merge-script/strategies/api-merge.ts (hunk 27 above)
Comment (josephperrott) [reply scripts/merge-script/strategies/api-merge.ts:149]: SGTM

File: scripts/merge-script/pull-request.ts [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+    return PullRequestFailure.claUnsigned();
+  }
+
+  const {data: {state}} =
+      await git.api.repos.getCombinedStatusForRef({...git.repoParams, ref: prData.head.sha});
+
+  const targetLabel = getTargetLabelFromPullRequest(config, labels);
+  if (targetLabel === null) {
+    return PullRequestFailure.noTargetLabel();
+  }
```
Comment (josephperrott) [inline scripts/merge-script/pull-request.ts:63]: Can we move the data request to after the target label check?  Just so we don't request the data and then fail before we even use it.
```suggestion
  const targetLabel = getTargetLabelFromPullRequest(config, labels);
  if (targetLabel === null) {
    return PullRequestFailure.noTargetLabel();
  }
  
  const {data: {state}} =
      await git.api.repos.getCombinedStatusForRef({...git.repoParams, ref: prData.head.sha});
```

Comment (devversion) [reply scripts/merge-script/pull-request.ts]: Yeah, fixed!
